### ISSUE TITLE
Truck moves backwards

### ISSUE THREAD

**Issue by thanif** [marshq/europilot#5]
While testing, he truck just moves in the backward direction and I am unable to understand why this is so.

**Comment by thanif** [comment]
uinput.ABS_X + (-32767, 32767, 0, 0),   # steering wheel
                uinput.ABS_Y + (0, 65535, 0, 0),   # clutch
                uinput.ABS_Z + (0, 65535, 0, 0),   # acceleration
                uinput.ABS_RZ + (0, 65535, 0, 0),  # break
                uinput.ABS_HAT0X + (-1, 1, 0, 0),
                uinput.ABS_HAT0Y + (-1, 1, 0, 0),

**Comment by thanif** [comment]
What is the reason for adding values like (0,65535, 0, 0) ,  (-1, 1, 0, 0), (-32767, 32767, 0, 0)

**Comment by oliver-im** [comment]
Hey thanif,

Please check which mode the gear is in. Also check the settings page in the game to see that you have properly mapped the joystick to the corresponding controls.

As for adding the values, they are added so that the joystick output values are normalized, so that they are more intuitive to the human reader. The values are probably specific to the Logitech G27 controller.

**Comment by thanif** [comment]
I trained the network on about 86000 frames at 20 FPS using "simple automatic" and changing the __init__ function in joystick.py as follows:
def __init__(self, name='DragonRise Inc. Generic USB Joystick', bustype=0x3,
                 vendor=0x79, product=0x6, version=0x110, events=None):
        if events is None:
            events = (
                # EV_KEY
                uinput.BTN_TRIGGER,
                uinput.BTN_THUMB,
                uinput.BTN_THUMB2,
                uinput.BTN_TOP,
                uinput.BTN_TOP2,
                uinput.BTN_PINKIE,
                uinput.BTN_BASE,
                uinput.BTN_BASE2,
                uinput.BTN_BASE3,
                uinput.BTN_BASE4,
                uinput.BTN_BASE5,
                uinput.BTN_BASE6,
                

                # EV_ABS
                uinput.ABS_X+ (-32767, 32767, 0, 0),
                uinput.ABS_Y+ (0, 65535, 0, 0), 
                uinput.ABS_Z+ (0, 65535, 0, 0), 
                uinput.ABS_RX+ (0, 65535, 0, 0), 
                uinput.ABS_RZ+ (0, 65535, 0, 0), 
                uinput.ABS_HAT0X+ (-1, 1, 0, 0), 
                uinput.ABS_HAT0Y+ (-1, 1, 0, 0),
            )

        self.device = uinput.Device(events,
                                    name=name,
                                    bustype=bustype,
                                    vendor=vendor,
                                    product=product,
                                    version=version)
While testing, the truck just goes in the straight direction using "keyboard+DragonRise Inc. Generic USB Joystick" mode, I can make the truck to turn using keyboard but this should not happen. Why this is so?
Some of the predicted values are: 450, 400, 470 etc.

**Comment by oliver-im** [comment]
Hey thanif,

1. Check if "'DragonRise Inc. Generic USB Joystick'" is shown when running
`cat /proc/bus/input/devices`

2. In the game, check if the driver is selected in the controllers settings page.

Provided that all the mappings are done right, the predicted values should turn the steering wheel inside the game.

**Comment by thanif** [comment]
The two conditions are met successfully, in the controllers page "keyboard+DragonRise Inc. Generic USB Joystick" is selected by default. You can see here.
![screenshot from 2017-10-09 17-13-58](https://user-images.githubusercontent.com/32266960/31337625-9840c7c2-ad15-11e7-99f6-f8bb5ff8405e.png)
Can you please tell me how can I verify my mappings.

**Comment by thanif** [comment]
I have got three values (0,32767,-32767) for straight, right and left movement and I divide these values by 655.35 and now the values are (0,-49.999237049,49.999237049) fed as labels while training but when I test the model the output before multiplying to 655.35 is like 0.966, 0.665 etc but it should be like 49.02, -49.25 and I am unable to understand why this is so.

**Comment by oliver-im** [comment]
Unfortunately I'm afraid I don't have a clear solution to the problem your facing. If you would be kind enough to post your workaround, it will be much appreciated.

**Comment by houwenbo87** [comment]
Hi thanif, I also encountered the same problem. Did you solve this problem?

**Comment by thanif** [comment]
I think that there might be a difference in the training methodology which I have used and now I am trying to train it using the data uploaded by @im9uri  at  https://drive.google.com/file/d/0B42sVbnSOCJ4bnZhWF80b0xUY28/view. This is because I found out that the prediction values of my model were forcing the truck to move backwards which I believe is caused as a result of some training problem and to confirm that I am going to use the data being uploaded instead of generating my own for training.

**Comment by thanif** [comment]
Thank you very much for releasing training data @im9uri . Now the truck moves as expected but there are still issues and I will ask as required. Hoping for a great adventure ahead as I further dive deep in to it.

**Comment by aryarohit07** [comment]
@thanif how did you resolve this? I am using the default implementation of the JoyStick provided and trained on the data uploaded by @im9uri . Do I have to do something (change gear?) in the settings? Any help would be much appreciated.
The truck moves backward even without emitting any values. Hence seems like some issue with instantiation.

**Comment by thanif** [comment]
Just before the start of game, we have the option to set controls like
"centred" or "inverted" for the play and the solution is to simply "centre"
the forward operation while "invert" all other ones like clutch or brake
etc.

Hope this helps.

**Comment by aryarohit07** [comment]
Thank you for a quick response. I tried the same:
Controller subtype -> Gamepad Joystick, Simple Automatic.

steering_axis -> Joy Wheel (centered)
Acceleration_axis -> Joy Clutch (inverted)
Brake_axis -> Joy Clutch (inverted)
Rest are NA

Truck's backward movement is solved but the steering doesn't seem to take any input from Joy Stick. Truck's steering is moving only by keyboard.

**Comment by thanif** [comment]
In my case, it worked by selecting the steering to "centred" and the other
two to "inverted and centred" and if it's not working then there might be
some issue with your code being used for loading weights or may be
mislabeled data or wrong mapping of the values.
